CYS: Show Patterns from PTK

gigitux opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This PR shows the patterns fetched from the PTK source. The user clicking on them can add in the Site.

Closes #47461 #47463 #47465 .

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Ensure that you have the last version of Gutenberg enabled
2. Make sure you have the WooCommerce beta tester plugin installed
3. Head over to Tools > WCA Test Helper > Features
4. Make sure you see the `pattern-toolkit-full-composability` on the list.
5. Enable it.
6. Head over to WooCommerce > Home > Customize your store.
7. Click on Design Your Homepage.
8. Click a category.
9. Click on different patterns.
10. Ensure that the patterns are added to the Site.
11. Click on a pattern.
12. Ensure that the tooltip is under the mouse pointer.
13. Save.
14. Ensure that the patterns are visible on the frontend.


<video src="https://github.com/woocommerce/woocommerce/assets/4463174/6e06aa2a-c2de-428e-9f94-901de7d9d04f" />



<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

CYS: Show Patterns from PTK

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

gigitux committed:
WIP

gigitux committed:
CYS: Refactor routing approach

gigitux committed:
Merge branch 'refactor/cys-route-system' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
fix type

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
remove focus logic

gigitux committed:
Merge branch 'refactor/cys-route-system' of github.com:woocommerce/woocommerce into refactor/cys-route-system

gigitux committed:
remove routekey prop

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into refactor/cys-route-system

gigitux committed:
Merge branch 'refactor/cys-route-system' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
fix scroll after insert pattern

gigitux committed:
pass onNavigateBackClick callback as prop

gigitux committed:
Merge branch 'refactor/cys-route-system' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
fix scroll

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
fix back arrow click

gigitux committed:
Merge branch '47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
fix css

gigitux committed:
fix tooltip position

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into refactor/cys-route-system

gigitux committed:
create dedicated CSS class

gigitux committed:
lint css

gigitux committed:
Merge branch 'refactor/cys-route-system' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
restore change

gigitux committed:
improve css class name

gigitux committed:
remove not necessary class

gigitux committed:
add github link to the todo

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
lint css

albarin commented:
@gigitux Thanks for working on this! A couple of comments from testing (I still need to check the code):

1. I noticed the `Want more patterns` section I added [here](https://github.com/woocommerce/woocommerce/pull/48095) is not showing, was that on purpose? I think it should show, right?
<img width="204" alt="Screenshot 2024-06-11 at 16 11 48" src="https://github.com/woocommerce/woocommerce/assets/186112/b409892e-7a78-4785-9096-d62559a0b08f">

2. Related to that, I had `Allow usage of WooCommerce to be tracked` disabled, so I didn't have any patterns imported from PTK, and because of that my `Intro` category showed empty:
<img width="687" alt="Screenshot 2024-06-11 at 16 14 51" src="https://github.com/woocommerce/woocommerce/assets/186112/dc351454-fce6-4f30-bc2c-e534ab3f8b88">
Should be prevent showing empty categories?

gigitux committed:
add want more pattern section

gigitux commented:
> 1. I noticed the `Want more patterns` section I added [here](https://github.com/woocommerce/woocommerce/pull/48095) is not showing, was that on purpose? I think it should show, right?

Great catch! I forgot to add them! Fixed with [daf534b](https://github.com/woocommerce/woocommerce/pull/48207/commits/daf534be50027bbdb64c5544f072842e26c9d986)

> Related to that, I had Allow usage of WooCommerce to be tracked disabled, so I didn't have any patterns imported from PTK, and because of that my Intro category showed empty:
Interesting question! If I'm not wrong for each category, we decided to keep some fallback bundled in the plugin to avoid this kind of issue, no? BTW, I can open an issue as a reminder to ensure that we double-check this use case.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

nefeline commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/pattern-screen/sidebar-pattern-screen.tsx:
```diff
@@ -0,0 +1,146 @@
+import { useEditorBlocks } from '../../hooks/use-editor-blocks';
+import { PATTERN_CATEGORIES } from './categories';
+
+export const SidebarPatternScreen = ( { category }: { category: string } ) => {
```
Do you mind clarifying why we aren't using Gutenberg's InserterSidebar component here? https://github.com/WordPress/gutenberg/blob/550cb6b67b544c72d124e8d6741e0d5577da9256/packages/editor/src/components/inserter-sidebar/index.js#L20C25-L20C40

nefeline commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/pattern-screen/categories.ts:
```diff
@@ -0,0 +1,27 @@
+import { __ } from '@wordpress/i18n';
+
+// @todo: Implement the right copy for the pattern categories: https://github.com/woocommerce/woocommerce/issues/48381
+export const PATTERN_CATEGORIES = {
```
Can we make a change here to ensure the categories are based on all patterns fetched via the PTK endpoint instead of having them hardcoded this way?

gigitux replied:
Given that this looks like a broader discussion (including some changes on the PKT API), what do you think if I open a dedicated issue?

nefeline replied:
> what do you think if I open a dedicated issue?

sounds good 👍

gigitux replied:
https://github.com/woocommerce/woocommerce/issues/48381

nefeline commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/navigation-extra-screen/sidebar-navigation-extra-screen.tsx:
```diff
@@ -0,0 +1,56 @@
+import './style.scss';
+import { PATTERN_CATEGORIES } from '../pattern-screen/categories';
+
+const getPatternCategoryByPath = ( path: string ) => {
```
Why aren't we basing this implementation on Gutenberg's [Inserter Sidebar](https://href.li/?https://github.com/WordPress/gutenberg/blob/550cb6b67b544c72d124e8d6741e0d5577da9256/packages/editor/src/components/inserter-sidebar/index.js#L90) and [InserterLibrary](https://github.com/WordPress/gutenberg/blob/15b8049af46b1849e96bc52c8d99386231bcc558/packages/block-editor/src/components/inserter/library.js#L15) ?

gigitux replied:
Yes! The `InsertSidebar` isn't a public component, but wraps the `InsertLibrary`. This latter, instead, doesn't allow us to hide the header and render only the pattern categories. Futhermore, even if we are able to do it, that component renders ALL the pattern categories that are registered; instead, we want to render only some of them.

nefeline replied:
WDYT about proposing upstream changes to support the usage of `InserterLibrary` in this particular case? There's a concern from the leadership about duplicating work and creating a "new editor" so I would recommend at least exploring this approach before proceeding with creating our dedicated component instead. Reference: pdnLyh-6dS-p2#comment-4563

gigitux replied:
Good observation! Thanks for pointing this out! 

> WDYT about proposing upstream changes to support the usage of InserterLibrary in this particular case?


IMHO, a tailored inserter with only specific categories isn't in the goal of Gutenberg: even if it is in "simple mode," by definition, Site Editor has to be extendible by 3rd party plugins, so I think that it is safe to assume that all the categories should always be visible.



Also, I don't see a lot of duplicate work here because this component is just a wrapper around the `BlockPatternList` + custom handlers.

nefeline reviewed (commented):
Thanks for your work here, @gigitux ! I just did my initial round here, focusing solely on code review in this first moment, left a few questions/suggestions.

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
remove not used import

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
fix logic after merge trunk

gigitux commented:
With [8559c05](https://github.com/woocommerce/woocommerce/pull/48207/commits/8559c059eaa904587e2b7915bed7f0762cbef293), I updated some logic about the interactive blocks given that the URL in case of the homepage could be: `/customize-store/assembler-hub/homepage/{pattern_category}` cc @albarin

gigitux committed:
CYS: fix tooltip position

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

gigitux committed:
trigger CI

gigitux committed:
Merge branch 'fix/cys-tooltip-position' of github.com:woocommerce/woocommerce into fix/cys-tooltip-position

gigitux committed:
Merge branch 'fix/cys-tooltip-position' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

WunderBart commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/navigation-extra-screen/sidebar-navigation-extra-screen.tsx:
```diff
@@ -0,0 +1,56 @@
+	const match = path.match( regex );
+
+	// If a match is found, return the captured group (everything after "/homepage/")
+	if ( match && match[ 1 ] ) {
+		return match[ 1 ];
+	}
+
+	const pattern_category = match && match[ 1 ];
```
Likely not intentional, as `pattern_category` will always be false in this context!

gigitux replied:
Nice catch! Fixed with: 8ec4ae631fe9f67998fa1d67bec72b6f1d2b0a67

WunderBart commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/navigation-extra-screen/sidebar-navigation-extra-screen.tsx:
```diff
@@ -0,0 +1,56 @@
+
+const getPatternCategoryByPath = ( path: string ) => {
+	// Define the regex pattern to match everything after "/homepage/"
+	const regex = /\/homepage\/(.*)/;
```
nit: If we want to be sure the path starts with `/homepage/`, we can validate with `path.startsWith( '/homepage/' )`.

gigitux replied:
We can't do it in this way because:
- the URL is: `admin.php?page=wc-admin&customizing=true&path=/customize-store/assembler-hub/homepage/${category}`
- I want to get the `category` variable.

WunderBart commented on plugins/woocommerce-admin/client/customize-store/assembler-hub/sidebar/navigation-extra-screen/sidebar-navigation-extra-screen.tsx:
```diff
@@ -0,0 +1,56 @@
+	const match = path.match( regex );
+
+	// If a match is found, return the captured group (everything after "/homepage/")
+	if ( match && match[ 1 ] ) {
```
nit: `match?.[1]`

gigitux replied:
Fixed with 316b8f0177b3e814b2796c5ef7fe99f2a7c50d34

gigitux committed:
Merge branch 'trunk' of github.com:woocommerce/woocommerce into 47461-cys-full-composability-ensure-pattern-categories-are-displayed-on-the-design-your-homepage-section

gigitux committed:
improve logic

gigitux committed:
use Optional Chaining
